server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldSyntheticWriterHelper.java

```diff
@@ -226,19 +225,27 @@ public boolean equals(Object obj) {
         }
     }
 
-    private final SortedSetDocValues dv;
+    interface SortedKeyedValues {
+        BytesRef next() throws IOException;
+    }
+
+    private final SortedKeyedValues sortedKeyedValues;
 
-    FlattenedFieldSyntheticWriterHelper(final SortedSetDocValues dv) {
-        this.dv = dv;
+    FlattenedFieldSyntheticWriterHelper(final SortedKeyedValues sortedKeyedValues) {
+        this.sortedKeyedValues = sortedKeyedValues;
     }
 
     void write(final XContentBuilder b) throws IOException {
-        KeyValue curr = new KeyValue(dv.lookupOrd(dv.nextOrd()));
+        KeyValue curr = new KeyValue(sortedKeyedValues.next());
         KeyValue prev = KeyValue.EMPTY;
         final List<String> values = new ArrayList<>();
         values.add(curr.value());
-        for (int i = 1; i < dv.docValueCount(); i++) {
-            KeyValue next = new KeyValue(dv.lookupOrd(dv.nextOrd()));
+        while (true) {
+            BytesRef nextValue = sortedKeyedValues.next();
+            if (nextValue == null) {
+                break;
```

Comment:
Nit: `for (BytesRef nextValue = sortedKeyedValues.next(); nextValue != null; nextValue = sortedKeyedValues.next()) {`